internal/stopwatch/stopwatch.go

```diff
@@ -0,0 +1,175 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package stopwatch
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// Stopwatch is a simple interface to measure time.
+// Stopwatch implementations are not thread-safe.
+type Stopwatch interface {
+	// Start starts the stopwatch and returns the start time.
+	Start() time.Time
+	// Remaining returns the remaining time before the stopwatch expires.
+	Remaining() time.Duration
+	// Expired returns true if the stopwatch has expired.
+	Expired() bool
+	// Stop stops the stopwatch and returns the time spent.
+	Stop() time.Duration
+	// Spent returns the time spent on the stopwatch.
+	Spent() time.Duration
+}
+
+type unlimitedStopwatch struct {
+	start time.Time
+	end   time.Time
+}
+
+// NewUnlimitedStopwatch returns a new unlimited stopwatch with no expiration.
+func NewUnlimitedStopwatch() Stopwatch {
+	return &unlimitedStopwatch{}
+}
+
+func (stopwatch *unlimitedStopwatch) Start() time.Time {
+	stopwatch.start = time.Now()
+	return stopwatch.start
+}
+
+func (stopwatch *unlimitedStopwatch) Remaining() time.Duration {
+	return 0
+}
+
+func (stopwatch *unlimitedStopwatch) Expired() bool {
+	return false
+}
+
+func (stopwatch *unlimitedStopwatch) Stop() time.Duration {
+	stopwatch.end = time.Now()
+	return stopwatch.Spent()
+}
+
+func (stopwatch *unlimitedStopwatch) Spent() time.Duration {
+	return stopwatch.end.Sub(stopwatch.start)
+}
+
+type timedStopwatch struct {
+	start    time.Time
+	duration time.Duration
+	end      time.Time
+	max      time.Time
```

Comment:
Would love some commentary to indicate what each field represents.

Also, `max` perhaps should be called `cap` or `limit` or `deadline`.